staking/slash/double-sign.go

```diff
@@ -1,52 +1,385 @@
 package slash
 
 import (
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+
 	"github.com/ethereum/go-ethereum/common"
-	"github.com/harmony-one/bls/ffi/go/bls"
+	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/harmony-one/harmony/block"
+	"github.com/harmony-one/harmony/common/denominations"
+	"github.com/harmony-one/harmony/consensus/votepower"
 	"github.com/harmony-one/harmony/core/state"
+	common2 "github.com/harmony-one/harmony/internal/common"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/numeric"
 	"github.com/harmony-one/harmony/shard"
+	staking "github.com/harmony-one/harmony/staking/types"
+	"github.com/pkg/errors"
+)
+
+const (
+	haveEnoughToPayOff               = 1
+	paidOffExact                     = 0
+	debtCollectionsRepoUndelegations = -1
+	validatorsOwnDel                 = 0
 )
 
+// invariant assumes snapshot, current can be rlp.EncodeToBytes
+func mustStayAboveZeroAfterDebtReduce(
```

Comment:
rename to "payDebtWithSanityCheck" or just "payDebt". Better use verb for a func that make changes.